File: python/p0001_two_sum/solutions/sorted_index_map_binary_search.py
# Time complexity: O(n log n)
# Space complexity: O(n)

from typing import List, Tuple


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        sorted_map = sorted((num, i) for i, num in enumerate(nums))

        def binary_search(left: int, right: int, search_val: int) -> int:
            # Return index of search_val in sorted_map
            while left <= right:
                mid = left + (right - left) // 2

                if sorted_map[mid][0] == search_val:
                    return mid
                elif sorted_map[mid][0] < search_val:
                    left = mid + 1
                else:
                    right = mid - 1
            
            return -1


        for i, (num, index) in enumerate(sorted_map):
            complement = target - num

            # Search only in the remaining list
            j = binary_search(i + 1, len(sorted_map) - 1, complement)

            if j != -1:
                return [index, sorted_map[j][1]]

File: rust/p0001_two_sum/src/sorted_index_map_binary_search.rs
// Time complexity: O(n log n)
// Space complexity: O(n)

pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        // Sort the numbers as (number, index) tuples to preserve original indices
        let mut sorted_map: Vec<(i32, usize)> =
            nums.iter().enumerate().map(|(i, &num)| (num, i)).collect();

        sorted_map.sort_by_key(|&(num, _)| num);

        // Return index of search_val in sorted_map
        fn binary_search(
            sorted_map: &[(i32, usize)], mut left: usize, mut right: usize, search_val: i32,
        ) -> Option<usize> {
            while left <= right {
                let mid = left + (right - left) / 2;

                if sorted_map[mid].0 == search_val {
                    return Some(mid);
                } else if sorted_map[mid].0 < search_val {
                    left = mid + 1
                } else {
                    right = mid - 1
                }
            }
            None
        }

        // Locate the complement
        for (i, &(num, index)) in sorted_map.iter().enumerate() {
            let complement = target - num;

            // Search only in the remaining vector slice
            if let Some(j) = binary_search(&sorted_map, i + 1, sorted_map.len() - 1, complement) {
                return vec![index as i32, sorted_map[j].1 as i32];
            }
        }

        unreachable!("This should be unreachable.");
    }
}
